### PR TITLE
Implement custom message class

### PR DESCRIPTION
**Pull request by andersjonsson**
Proof-of-concept-branch with additional performance work

Right now this is only used for reading the Request message. If it is going to be used for the Response message I have to figure out how this relates to CustomMessage. Should I merge the two implementations? Probably...

This branch uses a custom implementation of Message to make working with it more efficient. I also had to implement a custom XmlReader since the one returned by `XDocument.GetReader()` doesn't support binary data.

All tests pass, but nasty edge case bugs are still possible.

**develop**
| Method    | LoopNum | Mean     | Error     | StdDev    | Gen0      | Allocated |
|---------- |-------- |---------:|----------:|----------:|----------:|----------:|
| EmptyTask | 100     | 1.456 ms | 0.0474 ms | 0.0123 ms |  261.7188 |   1.05 MB |
| Echo      | 100     | 8.745 ms | 0.6351 ms | 0.1649 ms | 1656.2500 |   6.64 MB |

**This branch**
| Method    | LoopNum | Mean     | Error     | StdDev    | Gen0      | Gen1    | Allocated |
|---------- |-------- |---------:|----------:|----------:|----------:|--------:|----------:|
| EmptyTask | 100     | 1.550 ms | 0.0965 ms | 0.1112 ms |  263.6719 |  1.9531 |   1.05 MB |
| Echo      | 100     | 7.110 ms | 0.0646 ms | 0.0718 ms | 1187.5000 | 31.2500 |   4.82 MB |

### CODE REVIEW

**Comment by andersjonsson** [comment]
@akshaybheda I know you've asked about performance before. Do you have a good scenario for doing some benchmarks on this? Seems to me that the perf is better but I'd like to be sure that it also applies to a real scenario.

**Comment by akshaybheda** [comment]
Let me find some time and do some testing around this

**Comment by akshaybheda** [comment]
This branch has slightly better performance than the latest version 1.2.0
Latest version

image

Your branch with changes
image

File: src/SoapCore/SoapEndpointMiddleware.cs [hunk 1]
```diff
@@ -449,7 +449,7 @@ private async Task<Message> ProcessMessage(Message requestMessage, SoapMessageEn
 		{
 			Message responseMessage;
 			var soapAction = HeadersHelper.GetSoapAction(httpContext, ref requestMessage);
-			requestMessage.Headers.Action = soapAction;
+			requestMessage.Headers.Action = soapAction.ToString();
```
Comment (akshaybheda) [inline src/SoapCore/SoapEndpointMiddleware.cs]: In the GetSoapAction method, you are already doing the ToString()

File: src/SoapCore/XDocumentXmlReader.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+using System.Linq;
+using System.Text;
+using System.Xml;
+using System.Xml.Linq;
+
+namespace SoapCore
+{
+	public class XDocumentXmlReader : XmlReader
+	{
+		private readonly XDocument _document;
```
Comment (akshaybheda) [inline src/SoapCore/XDocumentXmlReader.cs]: never used

File: src/SoapCore/XDocumentXmlReader.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Xml;
+using System.Xml.Linq;
+
+namespace SoapCore
+{
+	public class XDocumentXmlReader : XmlReader
```
Comment (akshaybheda) [inline src/SoapCore/XDocumentXmlReader.cs:11]: you can also implement IDisposable to dispose the streams

File: src/SoapCore/Extensibility/AuthorizeOperationMessageProcessor.cs [hunk 4]
```diff
@@ -43,7 +43,7 @@ public AuthorizeOperationMessageProcessor(Dictionary<string, Type> pathAndTypes,
 		public async Task<Message> ProcessMessage(Message requestMessage, HttpContext httpContext, Func<Message, Task<Message>> next)
 		{
 			var soapAction = HeadersHelper.GetSoapAction(httpContext, ref requestMessage);
-			requestMessage.Headers.Action = soapAction;
+			requestMessage.Headers.Action = soapAction.ToString();
```
Comment (akshaybheda) [inline src/SoapCore/Extensibility/AuthorizeOperationMessageProcessor.cs]: I don't think this is needed

File: src/SoapCore/MessageEncoder/SoapMessageEncoder.cs [hunk 5]
```diff
@@ -136,19 +136,7 @@ public async Task<Message> ReadMessageAsync(Stream stream, int maxSizeOfHeaders,
...
-			}
-			else
-			{
-				ms = new MemoryStream();
-				await stream.CopyToAsync(ms);
-				ms.Seek(0, SeekOrigin.Begin);
-			}
-
-			XmlReader reader;
+			Message message;
```
Comment (akshaybheda) [inline src/SoapCore/MessageEncoder/SoapMessageEncoder.cs:139]: I don't think you need async Task<Message>, there are no awaits here, you could just return Message and update the callers

File: src/SoapCore/MessageEncoder/SoapMessageEncoder.cs [hunk 6]
```diff
@@ -162,17 +150,23 @@ public async Task<Message> ReadMessageAsync(Stream stream, int maxSizeOfHeaders,
 
 			if (supportXmlDictionaryReader)
 			{
-				reader = XmlDictionaryReader.CreateTextReader(ms, readEncoding, ReaderQuotas, dictionaryReader => { });
+				using (var xdReader = XmlDictionaryReader.CreateTextReader(stream, readEncoding, ReaderQuotas, dictionaryReader => { }))
```
Comment (akshaybheda) [inline src/SoapCore/MessageEncoder/SoapMessageEncoder.cs:153]: nitpick: you can simplify the using

**Review by akshaybheda — CHANGES_REQUESTED** [review]
Did look at some of the changes and added some comments

**Comment by andersjonsson** [comment]
Thank you for taking the time.
I was hoping for a more clear performance win, but I'll take it :)

I also managed to get ParsedMessage working for the response, which resulted in higher memory usage and no speed improvement, so I'll give up on that idea.

I also tried rewriting the DefaultOperationInvoker to use expression trees, but that didn't result in a noticeable win. Maybe with more complex operations... I put that implementation in a gist, in case you want to try it out for your scenario
https://gist.github.com/andersjonsson/004925c250695a44ab7c2f0637dfd465

Rewriting the invocation code to use source generators would most likely be a big benefit, but that would be a big project and would not work for netstandard 2.0.

File: src/SoapCore/XDocumentXmlReader.cs (hunk 3 above)
Comment (andersjonsson) [reply src/SoapCore/XDocumentXmlReader.cs:11]: XmlReader already implements IDisposable but I added an overload to dispose the stream

File: src/SoapCore/MessageEncoder/SoapMessageEncoder.cs (hunk 5 above)
Comment (andersjonsson) [reply src/SoapCore/MessageEncoder/SoapMessageEncoder.cs:139]: Fixed

File: src/SoapCore/SoapEndpointMiddleware.cs (hunk 1 above)
Comment (andersjonsson) [reply src/SoapCore/SoapEndpointMiddleware.cs]: Good catch! Fixed

**Review by akshaybheda — APPROVED** [review]
Look good to me
